scripts/tinkerbell2answers.sh

```diff
@@ -0,0 +1,179 @@
+#!/bin/bash
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+# Copyright 2021 Joyent, Inc.
+
+# This script will take a metadata file provided as a boot module from
+# Equinix Tinkerbell (e.g., Equinix Metal) and generate a suitable answers.json
+# file. If the metadata is provided as a bood module, prompt-config.sh will
+# call this script and then use the generated answers file for unattended
+# install.
+
+# The metadata file gets passed in on the command line by prompt-config.sh
+md="$1"
+
+# shellcheck disable=2154
+if [[ -n "$TRACE" ]]; then
+    # BASHSTYLED
+    export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+    set -o xtrace
+fi
+set -o errexit
+set -o pipefail
+
+if ! [[ -f $md ]]; then
+    printf 'No tinkerbell metadata\n'
+    exit
+fi
+
+# This is a short hand sanity test to make sure the metadata file is valid
+# JSON, and likely actually came from tinkerbell.
+if ! json -f "$md" iqn | grep -q iqn ; then
+    printf 'Metadata file is not valid\n' >&2
+    exit 1
+fi
+
+# Number of IPs by prefix-length
+# num_ips[24] == 256
+num_ips=(
+    4294967296
+    2147483648
+    1073741824
+    536870912
+    268435456
+    134217728
+    67108864
+    33554432
+    16777216
+    8388608
+    4194304
+    2097152
+    1048576
+    524288
+    262144
+    131072
+    65536
+    32768
+    16384
+    8192
+    4096
+    2048
+    1024
+    512
+    256
+    128
+    64
+    32
+    16
+    8
+    4
+    2
+    1
+)
+
+# Return a value or default if it doesn't exist
+function get_default {
+    local val
+    val=$(get_value "$1")
+    if (( ${#val} < 1 )); then
+        echo "$2"
+        return
+    fi
+    echo "$val"
+}
+
+function get_value {
+    json -f "$md" "$@"
+}
+
+function random_octet {
+    echo $(( RANDOM % 255))
+}
+
+# These will always be present
+datacenter_name="$(get_value facility)"
+admin_nic="$(get_value network.interfaces.1.mac)"
+external_nic="$(get_value network.interfaces.0.mac)"
+addresses="$(get_value network.addresses)"
+
+# Derived values
+IFS=: read -a a_mac <<< "$admin_nic"
+admin_net=$(printf '10.%d.%d' "0x${a_mac[4]}" "0x${a_mac[5]}")
```

Comment:
Here we're deterministically generating the 2nd & 3rd octet for the admin network based on the final two octets of the admin nic mac address.

I'd like to solicit opinions on this. Should I make this overridable via `customdata`?

Reply:
If you don't imagine a /24 is too small, this is a great idea.  I also see you reserve .1 to .9 below for presumably manual-configuration escapes. And IMHO yes, you should allow an override via customdata.

Reply:
Ok, I took your advice on this which necessitated switching it to node.js.

Now it will generate a /22 network based on the mac suffix.